### DISCUSSION TITLE
ZSH functions defined in completion script evaluate lazily

### DISCUSSION THREAD

**Discussion by jjmark15** [zellij-org/zellij#1860] [General]
Hi, I added the ZSH completion script generated by Zellij to my fpath but I have found that until I `tab` for completion with the `zellij` command, none of the ZSH functions e.g. `zef` are evaluated and available in my shell. Is this expected behaviour?

**Comment by imsnif** [comment]
This is not expected behaviour. :) Truth be told, I'm not a big zsh user so I wouldn't know where to put the aliases. Do you have any tips?

**Reply by jjmark15** [reply]
To be honest I am no expert on ZSH, I'm afraid I probably know less than you

**Comment by agluck91** [comment]
I know this thread is old, but I was having the same issue. This was driving me crazy, it was so illusive and hard to figure out what was going on. 
What I ended up doing was I pull the functions out and added them to my custom functions list and those are sourced in my .zshrc something like: `source $CONFIG/fucntions/zellij.zsh` .
Then the autocomplete works fine and the functions were always available too. 

@imsnif Maybe the generate completions function needs to not include the alias and maybe those need to get sourced a different way. I would love for a ZSH expert to correct me on here. I am not the smartest guy. I spent way too long trying to figure out why my autocomplete never loaded from my fpath, when it actually did, the functions just never loaded because the script never gets executed, even with auto-load I was still never able to get these functions to surface right. 

Hopefully this hack works for someone else in the future that is having this problem.

**Reply by agluck91** [reply]
fsh is starting to look more appealing in a lot of ways.

**Reply by cristiand391** [reply]
I explored this while working on zsh completion for [oclif](https://oclif.io/)  a few months ago and AFAIK you can't load "normal" functions in autocomplete files.

Autocomplete files start with `#compdef <bin>`  and define a `_<bin>` function that gets loaded when you do `<bin><TAB>`,  example opening a new shell:

I put zellij comp. file in `/home/cristiand391/.config/zsh/comp/_zellij`, when you run `which _zellij` you can see that it hasn't been loaded yet. Next I type `zellij ` and press tab to trigger completions (the `_zellij` func in the completion file gets loaded in memory) then Ctrl C. If I run `which _zellij` again now I see the completion function is in memory.

![image](https://github.com/zellij-org/zellij/assets/6853656/9388b1dd-3c97-40ad-ae70-1da65fd0466f)


the completion file generated by `zellij setup --generate-completion zsh` puts the run/edit functions at the end of file but since this is a comp def file zsh never loads them:
![image](https://github.com/zellij-org/zellij/assets/6853656/52d05173-82c4-452a-9c91-063f75dce5e7)


Unless there's some option I'm not aware of/hack, I think is better to remove them and put them under a section in the readme to manually add them to the zshrc file.

zsh docs:
https://zsh.sourceforge.io/Doc/Release/Completion-System.html#Autoloaded-files
https://zsh.sourceforge.io/Doc/Release/Functions.html
